### DISCUSSION TITLE
Retrieve first and last date in billable timesheet entries (custom invoice templates)

### DISCUSSION THREAD

**Discussion by lwndp** [kimai/kimai#3593] [Questions & Answers]
Hey! I'm trying to find a way to display the period of billing ("Leistungszeitraum") on an invoice. The start date should be the oldest date in the to-be-billed timesheet and the end date should be the latest one of them. The dates should be derived from billable timesheet entries that have not yet been exported in another invoice.

Results should look something like that:

Leistungszeitraum: 03.05.2022 - 05.07.2022

Does anyone know how to do that?

Thanks a lot for any pointers / help!!!

**Comment by BeckeBauer** [comment]
I've been working with `${query.begin}` and `${query.end}` variables in the invoice template which means that it has to match your query when creating the invoice. Wished there were an easier way...

**Reply by lwndp** [reply]
Thanks for the reply! I've tried that but it's not an optimal solution to my problem, as it could lead to timesheet entries not getting billed if the next query starts at a later date. Generally, a confusing way of doing it. I wish the variable was independent from the query and the template would just look up the oldest and youngest date in the retrieved timesheet entries.
Anyway, thanks for your help @BeckeBauer !

**Comment by Sarmingsteiner** [comment]
So, this cool feature was added in [Kimai 1.28
](https://github.com/kevinpapst/kimai2/releases/tag/1.28.0) with #3594.
Has anyone already found out which variables to add in the invoice templates?
And is it possible to show just month and year, e.g. "Leistungszeitraum: 05/2022 - 07/2022"?

**Reply by kevinpapst** [reply]
Documents:
`Leistungszeitraum ${invoice.first} - ${invoice.last}`

Twig:
`Leistungszeitraum {{invoice['invoice.first']}} - {{invoice['invoice.last']}}`

**Reply by BeckeBauer** [reply]
I tried `Leistungszeitraum ${invoice.first} - ${invoice.last}` with a word template and it shows the full date of the first and last entry, e.g. `Leistungszeitraum 05.10.2022 - 08.10.2022` (when the first timesheet dates Oct 5 and the last Oct 8). Is there any additional variable to skip the day so that only the month(s) is(are) shown?

**Reply by kevinpapst** [reply]
No.

**Comment by BeckeBauer** [comment]
great!!! I will check it out tomorrow...

**Comment by lwndp** [comment]
Awesome, thank you @kevinpapst ! 🙏🏼

**Heads up:**
Only works with specifying date_short in twig files for me.
Like so: {{invoice.first|date_short}} - {{invoice.last|date_short}}

**Reply by kevinpapst** [reply]
This is already a formatted string, which can be displayed as is.

**Reply by lwndp** [reply]
Just tested with invoice previews and only got one date, the upper bound, for both invoice.first & invoice.last.
Anyone else can reproduce? I'll file a bug report if so.

**Reply by lwndp** [reply]
> This is already a formatted string, which can be displayed as is.

Without specifying nothing gets printed though.

**Reply by kevinpapst** [reply]
You are likely receiving an invalid date, because you are applying a filter on a string. 
Fallback of the date filter is the current date,

**Reply by lwndp** [reply]
Maybe I wasn't clear earlier.
This "{{invoice.first}} - {{invoice.last}}" will print "-"

**Reply by kevinpapst** [reply]
Maybe I am not clear:
It is Saturday evening, 1.28 was released less than an hour ago and you feel a bit pushy, aren't you?

Test with the debug template, which is linked in the docs. Copy & Paste the template, show me the result.

<img width="421" alt="Bildschirmfoto 2022-10-29 um 20 59 08" src="https://user-images.githubusercontent.com/533162/198848402-e41e1eb6-0a17-4ade-8966-8ba998e497fa.png">

<img width="560" alt="Bildschirmfoto 2022-10-29 um 20 58 56" src="https://user-images.githubusercontent.com/533162/198848394-6f5f92d4-99b4-4173-a9a3-471d2a911d5e.png">

**Reply by lwndp** [reply]
I'm sorry if I came across as pushy. That was in no way my intention. Also, I did not expect an immediate reply - I wouldn't on a Saturday evening, nor during the week.

With a bit of back and forth I could eventually troubleshoot the problem and make it work. 

Sorry again for the misunderstanding!

**Reply by kevinpapst** [reply]
All good. Glad you found the issue (using a not existing variable leads to null, leads to the current date when using the date_short filter).
